File: 网易2019实习生招聘编程题集合/牛牛的背包问题.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long LL;

int main()
{
    for (LL n, w; cin >> n >> w; ) {
        vector<LL> v;
        for (int i = 0, x; i < n; cin >> x, v.push_back(x), ++i) {}

        LL n1 = n >> 1, n2 = n - n1;

        vector<LL> arr;
        for (int i = 0; i < (1 << n1); i++) {
            LL sum = 0;
            for (int bits = 0; bits < n1; bits++)
                sum += (i & (1 << bits)) ? v[bits] : 0;
            arr.push_back(sum);
        }
        sort(arr.begin(), arr.end());

        LL ans = 0;
        for (int i = 0; i < (1 << n2); i++) {
            LL sum = 0;
            for (int bits = 0; bits < n2; bits++)
                sum += (i & (1 << bits)) ? v[n1 + bits] : 0;
            ans += lower_bound(arr.begin(), arr.end(), w - sum + 1) - arr.begin();
        }
        cout << ans << endl;
    }
    return 0;
}

File: 网易2019实习生招聘编程题集合/牛牛的背包问题.java
import java.util.*;

public class Main {

	public static Scanner cin = new Scanner(System.in);

	public static int lower_bound(long[] arr, int begin, int end, long tag) {
		while (begin < end) {
			int mid = (begin + end) / 2;
			if (arr[mid] < tag)
				begin = mid + 1;
			else
				end = mid;
		}
		return begin;
	}

	public static void main(String[] args) {
		while (cin.hasNext()) {
			long n = cin.nextLong(), w = cin.nextLong();
			long[] v = new long[(int) n];
			for (int i = 0; i < n; v[i++] = cin.nextLong()) {
			}
			long n1 = n >> 1, n2 = n - n1;
			long[] arr = new long[1 << n1];
			for (int i = 0; i < (1 << n1); i++) {
				long sum = 0;
				for (int bits = 0; bits < n1; bits++)
					sum += (long) (((i & (1 << bits)) != 0) ? v[bits] : 0);
				arr[i] = sum;
			}
			Arrays.sort(arr);

			long ans = 0;
			for (int i = 0; i < (1 << n2); i++) {
				long sum = 0;
				for (int bits = 0; bits < n2; bits++)
					sum += (long) (((i & (1 << bits)) != 0) ? v[(int) n1 + bits]
							: 0);
				ans += lower_bound(arr, 0, arr.length, w - sum + 1);
			}
			System.out.println(ans);
		}
	}
}
